daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -232,19 +345,33 @@ class VariableMap private(vTable: Map[GlobalQName, MStackOf[VariableInstance]])
    */
   def readVariable(vrd: VariableRuntimeData, referringContext: ThrowsSDE, maybePstate: Maybe[ParseOrUnparseState]): DataValuePrimitive = {
     val varQName = vrd.globalQName
-    val stack = vTable.get(varQName)
-    if (stack.isDefined) {
-      val variable = stack.get.top
-      variable.state match {
-        case VariableRead if (variable.value.isDefined) => variable.value.getNonNullable
-        case VariableDefined | VariableSet if (variable.value.isDefined) => {
+    val abuf = vTable.get(varQName)
+    if (abuf.isDefined) {
+      val variable = abuf.get.last
+      (variable.state, variable.value.isDefined, variable.defaultValueExpr.isDefined) match {
+        case (VariableRead, true, _) => variable.value.getNonNullable
+        case (VariableDefined | VariableSet, true, _) => {
           if (maybePstate.isDefined && maybePstate.get.isInstanceOf[PState])
             maybePstate.get.asInstanceOf[PState].markVariableRead(vrd)
 
           variable.setState(VariableRead)
           variable.value.getNonNullable
         }
-        case _ => throw new VariableHasNoValue(varQName, vrd)
+        case (VariableDefined, false, true) => {
+          Assert.invariant(maybePstate.isDefined)
+          variable.setState(VariableInProcess)
+          val pstate = maybePstate.get
+          val res = DataValue.unsafeFromAnyRef(variable.defaultValueExpr.get.evaluate(pstate))
+
+          // Need to update the variable's value with the result of the
+          // expression
+          variable.setState(VariableRead)
+          variable.setValue(res)
+
+          res
```

Comment:
Is this correct? This is evaluating the defaultValue and setting the state when the variable is read. But the spec says this about newVariableInstance:

> If a default value is specified this initial value of the instance will be created when the instance is created

So it seems this evaluation should not happen during read, but when the instance of the variable is created. And then that default could be overridden with a setVariableCal.

Looks like you do evaluate the default expression below in newVariableInstance, so maybe this just needs to use what that sets?

Reply:
So this case here is really only for the case where we have a defineVariable with a default value expression that references another variable with a default value expression.  So, what happens is when we call the forceExpressoinEvaluations at the start of parsing, it evaluates the top level variable default value expression, which then calls readVariable for the variable that it references, which then needs it's expression evaluated at that time.  I'll add a comment to this case explaining what is happening when this case is hit.

This isn't an issue for newVariableInstance, as we do no need to postpone the evaluation of expressions because the VariableMap alreadys exists by the time we would run into any newVariableInstance expressions.

Reply:
Ah, makes sense. A comment explaining this would definitely be useful.